src/middleware.ts

```diff
@@ -258,36 +263,63 @@ export const persist = <S extends State>(
     void setItem()
   }
 
+  const makeThenable = (value: any) => {
+    if (value instanceof Promise) {
+      return value
+    } else {
+      return {
+        then: (callback) => callback(value),
+        // this is a fake function just to avoid catch is not a function error
+        catch: () => undefined,
+      }
+    }
+  }
   // rehydrate initial state with existing stored state
-  ;(async () => {
+  ;(() => {
     const postRehydrationCallback = onRehydrateStorage?.(get()) || undefined
 
-    try {
-      const storageValue = await storage.getItem(name)
-
-      if (storageValue) {
-        const deserializedStorageValue = await deserialize(storageValue)
-
-        // if versions mismatch, run migration
-        if (deserializedStorageValue.version !== version) {
-          const migratedState = await migrate?.(
-            deserializedStorageValue.state,
-            deserializedStorageValue.version
-          )
-          if (migratedState) {
-            set(migratedState)
-            await setItem()
+    let storageValuePending = makeThenable(storage.getItem(name))
+    storageValuePending
+      .then((storageValue) => {
+        if (storageValue) {
+          return makeThenable(deserialize(storageValue))
+        } else {
+          return makeThenable(null)
+        }
+      })
+      .then((deserializedStorageValue) => {
+        if (deserializedStorageValue) {
+          if (deserializedStorageValue.version !== version) {
+            return makeThenable(
+              migrate?.(
+                deserializedStorageValue.state,
+                deserializedStorageValue.version
+              )
+            )
+          } else {
+            set(deserializedStorageValue.state)
+            return makeThenable(null)
           }
         } else {
-          set(deserializedStorageValue.state)
+          return makeThenable(null)
         }
-      }
-    } catch (e) {
-      postRehydrationCallback?.(undefined, e)
-      return
-    }
-
-    postRehydrationCallback?.(get(), undefined)
+      })
+      .then((migratedState) => {
+        if (migratedState) {
+          set(migratedState)
+          return makeThenable(setItem())
+        } else {
+          return makeThenable(null)
+        }
+      })
+      .then(() => {
+        postRehydrationCallback?.(get(), undefined)
+        return makeThenable(null)
+      })
+      // this handler will only catch promise related errors
+      .catch((e) => {
```

Comment:
This `catch` handles only errors thrown in async mode. For sync mode the old `try...catch` needs to be reintroduced, meaning we need both `catch()` and `try...catch`.